### PR TITLE
Load environment fle in db tasks

### LINKED ISSUE #30717: Encrypted secrets not being read in some rake tasks
### Steps to reproduce
Try to perform `RAILS_ENV=production rails db:drop` with correctly setup of secrets.yml.enc with 'secret_key_base' variable. To handy error indication also 'devise' gem is needed, so it will raise error if no 'secret_key_base' variable was read. To read the variable from new secrets store, add ``config.secret_key = Rails.application.secrets.devise_secret_key if Rails.env.production?`` to devise.rb.
``config.read_encrypted_secrets = true`` is set in production.rb.
PS It is not only devise-related problem, some strange errors with the rest of variables happens also.
### Expected behavior
Successful database dropping (for sure with correct credentials are in place).
### Actual behavior
`Devise.secret_key was not set. Please add the following to your Devise initializer`. 
Some other rake tasks still works fine (db:migrate etc.)
There is no such problem in RAILS_ENV=development.
### Way to solve
If you move `config.read_encrypted_secrets = true` from `environments\production.rb` to `application.rb`, looks like every errors related to unread secrets are gone. For some reason config files in different environments are loaded in different orders, and probably in production a lot of them are loaded before ``config.read_encrypted_secrets = true`` will be hit.
### System configuration
**Rails version**:
5.1.3
**Ruby version**:
ruby 2.4.2p198 
Windows 7

### PR DESCRIPTION
**Pull request by y-yagi**
Currently the environment is not loaded in some db tasks.
Therefore, if use encrypted secrets values in `database.yml`, `read_encrypted_secrets` will not be true, so the value can not be used correctly.

To fix this, added `environment` task to the dependent task.
Fixes #30717

### CODE REVIEW

File: activerecord/lib/active_record/railties/databases.rake [hunk 1]
```diff
@@ -25,7 +25,7 @@ db_namespace = namespace :db do
   end
 
   desc "Creates the database from DATABASE_URL or config/database.yml for the current RAILS_ENV (use db:create:all to create all databases in the config). Without RAILS_ENV or when RAILS_ENV is development, it defaults to creating the development and test databases."
-  task create: [:load_config] do
+  task create: [:environment, :load_config] do
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/railties/databases.rake]: I think we need to change `load_config` to have `environment` as dependency. the `db:create:all` task now is not loading `environment` for example.

Comment (y-yagi) [reply activerecord/lib/active_record/railties/databases.rake]: Thanks for your comment. I agree that adding a dependency of `environment` to `load_config` is a better approach. Updated.

**Review by rafaelfranca — APPROVED** [review]
Feel free to merge it

**Comment by y-yagi** [comment]
Thanks! Merged :smile:

**Comment by jeremy** [comment]
🎉🎉🎉🎉🎉

**Comment by jeremy** [comment]
Noticed that this now causes `db:create` to fail if a schema dump is present. We load the environment which does a schema cache version check, but checking `ActiveRecord::Migrator.current_version` will always fail since the db doesn't yet exist.

https://github.com/rails/rails/blob/d041a1dcbaced9f44ed976a48312c9f492fffe5e/activerecord/lib/active_record/railtie.rb#L89-L101

To work around it, delete the schema cache dump. To fix, we'd need to skip the schema cache load if the db doesn't exist.

**Comment by y-yagi** [comment]
Thank you for the report! I will try to fix.

**Comment by eileencodes** [comment]
Hey @y-yagi, @tenderlove and I were working on Active Record and noticed an issue with this PR that is revealed by the test you wrote.

This PR ensures that the environment configs are loaded before the database.yml - but the database.yml cannot rely on the config the way your test demonstrates.

If you set a `config.database = "myapp_test"` in the test config and then set `config.database="myapp_development"` in the development config only the development value will be filled in. This is because the database.yml file can only be evaluated in one context. So if you try to name the test database something different in your test config and run rake tasks only the development config will be created and then the tests cannot run because the db won't exist.

Because the database.yml file contains information about more than one environment BUT only one environment can be loaded at once it has some serious side-effects depending on the environment it's run in. Basically the database.yml cannot depend on the environment configuration settings unless they are all the same (so it should be set in application.yml and not the environment). 

The one thing I'm unsure about is whether the test you wrote is to just provide an example of how the database can depend on config options or rather if it requires a dependency on environment?

**Comment by tenderlove** [comment]
Yes, the fundamental problem is that `database.yml` contains configuration information for all environments, but we only load one file from `config/environments` (like `config/environments/test.rb`). It doesn't make sense that the `database.yml` file would depend on something that is in `config/environments/test.rb`.

**Comment by matthewd** [comment]
@eileencodes @tenderlove there's only one special case (`test` for `development`) where we attempt to load a database configuration for an environment other than the one that's currently running, right? Maybe it'd make life simpler to just stop doing that, and spawn a subprocess in the right environment instead?

Secrets configuration depends on `config/environments/production.rb`; it's expected (desired) behaviour that a secrets-using DB config won't be complete if it's loaded in the wrong (non-secrets-supplying) environment.

**Comment by y-yagi** [comment]
The issue that required this fix that the DB tasks do not work properly when using encrypted secret.   

For example, if used encrypted secrets to manage the secret key of devise, need to read encrypted secrets before executing task even if do not use encrypted secrets in `database.yml`. 
However, in order for encrypted secrets to be read, must explicitly specify `true` for `read_encrypted_secrets`. 
Since we recommended that `read_encrypted_secrets` be defined in the environment file, we needed to read the environment file before executing the task. 

As a result, this correspondence was done to ensure that encrypted secrets are loaded, not the assumption of specifying database. 
It seems that the test was confusing due to bad, I am sorry.

**Comment by akostadinov** [comment]
This change unfortunately causes the whole application to load just to get secret usage. This happens by means of `Rails.application.initialize!` inside `environment.rb`.

Then a lot of gems can trigger a database access and `db:create` can't work anymore. As an example such gems like `thinking-sphinx` or observers. Then the task fails beause connection fails before database was created. Talking about MySQL where database name is in connection URL.

```
ActiveRecord::NoDatabaseError: Unknown database 'myapp_development'
```

In my opinion, if the whole problem was to know whether encrypted secrets should be used, then a solution to detect that in another way than loading the whole app should be used.

I've read so many threads about `db:create` failing and people resorting to direct database creation with native tools. I found this change only by thoroughly comparing execution flow of Rails 5.1.7 and Rails 5.2.7

**Comment by stuarthannig** [comment]
Upgrade fix by bug introduced by this.
Add `Rake::Task["db:load_config"].clear_prerequisites` to `Rakefile`

Src: https://stackoverflow.com/a/72191416

**Pull request merged**